pkg/virt-controller/watch/topology/nodetopologyupdater.go

```diff
@@ -0,0 +1,121 @@
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/client-go/log"
+)
+
+type NodeTopologyUpdater interface {
+	Run(interval time.Duration, stopChan <-chan struct{})
+}
+
+type nodeTopologyUpdater struct {
+	nodeStore cache.Store
+	hinter    Hinter
+	client    kubecli.KubevirtClient
+}
+
+type stats struct {
+	updated int
+	skipped int
+	error   int
+}
+
+func (n *nodeTopologyUpdater) Run(interval time.Duration, stopChan <-chan struct{}) {
+	wait.JitterUntil(func() {
+		requiredFrequencies := n.requiredFrequencies()
+		nodes := FilterNodesFromCache(n.nodeStore.List(),
+			HasInvTSCFrequency,
+		)
+		stats := &stats{}
+		for _, node := range nodes {
+			nodeCopy, err := calculateNodeLabelChanges(node, requiredFrequencies)
+			if err != nil {
+				stats.error++
+				log.DefaultLogger().Object(node).Reason(err).Error("Could not calculate TSC frequencies for node")
+				continue
+			}
+			if !reflect.DeepEqual(node.Labels, nodeCopy.Labels) {
+				if err := patchNode(n.client, node, nodeCopy); err != nil {
+					stats.error++
+					log.DefaultLogger().Object(node).Reason(err).Error("Could not patch TSC frequencies for node")
+					continue
+				}
+				stats.updated++
+			} else {
+				stats.skipped++
+			}
+		}
+		log.DefaultLogger().Infof("TSC Freqency node update status: %d updated, %d skipped, %d errors", stats.updated, stats.skipped, stats.error)
+	}, interval, 1.2, true, stopChan)
+}
+
+func patchNode(client kubecli.KubevirtClient, original *v1.Node, modified *v1.Node) error {
+	originalBytes, err := json.Marshal(original)
+	if err != nil {
+		return fmt.Errorf("could not serialize original object: %v", err)
+	}
+	modifiedBytes, err := json.Marshal(modified)
+	if err != nil {
+		return fmt.Errorf("could not serialize modified object: %v", err)
+	}
+	patch, err := strategicpatch.CreateTwoWayMergePatch(originalBytes, modifiedBytes, v1.Node{})
+	if err != nil {
+		return fmt.Errorf("could not create merge patch: %v", err)
+	}
+	if _, err := client.CoreV1().Nodes().Patch(context.Background(), original.Name, types.StrategicMergePatchType, patch, v12.PatchOptions{}); err != nil {
+		return fmt.Errorf("could not patch the node: %v", err)
+	}
+	return nil
+}
+
+func calculateNodeLabelChanges(original *v1.Node, requiredFrequencies []int64) (modified *v1.Node, err error) {
+	nodeFreq, scalable, err := TSCFrequencyFromNode(original)
+	if err != nil {
+		log.DefaultLogger().Reason(err).Object(original).Error("Can't determine TSC frequency of the original")
+		return nil, err
+	}
+	freqsOnNode := TSCFrequenciesOnNode(original)
+	toAdd, toRemove := CalculateTSCLabelDiff(requiredFrequencies, freqsOnNode, nodeFreq, scalable)
+	toAddLabels := ToTSCSchedulableLabels(toAdd)
+	toRemoveLabels := ToTSCSchedulableLabels(toRemove)
```

Comment:
this might be mostly theoretical, but i think there is a race condition here with this label removal.

if a vmi doesn't have topology hints assigned it could grab something from the node informer cache for scheduling and set it on the VMI while this node topology updater is removing it.  There's a small gap due to the node cache lagging behind that could result in the VMI getting a frequency that that doesn't have a matching label on a node anymore.

it might not be practical to close this gap.  It's just something i noticed.

Reply:
That should not be an issue. The worst thing that would happen is that the VMI can't be scheduled for 30 seconds, because on the reevaluation the topologyupdater would see the correct values on the VMI and would re-add the labels where possible.

Reply:
+1